### PR TITLE
Gallery app

### PR DESCRIPTION
**Pull request by yannickulrich**
# Gallery application
With this application you can add image and text files to the watches FS and view them on the watch (#1237). You can have multiple files on the watch by copying them into the directory `/gallery`. This would allow for example to have QR codes (or other barcodes such as Aztec codes which are often used for airline tickets) on the watch by creating them in the companion app (related #181). Further, you can have text files on the watch that will just be shown.

## Usage
You first need to make the directory `/gallery`. Using eg. `itctl`
```sh
$ itctl fs mkdir /gallery                          # ensure folder exists
```
Files with the suffix `.bin` are treated as images and passed to lvgl for rendering. Files with suffix `.txt` will be shown as text files.

Image files need to have the correct size (max 240x240) and be converted using `lv_img_conv` and can then be uploaded using eg. `itctl`
```sh
$ lv_img_conv code.png --output-file code.bin \
        --color-format CF_TRUE_COLOR \
        --output-format bin \
        --binary-format ARGB8565_RBSWAP            # convert image
$ itctl fs write code.bin /gallery/code.bin        # upload image
```
You can navigate between the files by swiping left and right. Long pressing the screen hides the file name and navigation bar.

## Screenshots
screencast
Text file
Image file

## Status
The app works well in the simulator and I have started testing it on a physical device. However, only test this on a sealed watch if you know what you are doing!

I'd appreciate any input but especially concerning memory leaks.

### CODE REVIEW

**Comment by yannickulrich** [comment]
Colours should be possible, yes. LVGL supports [re-colouring](https://docs.lvgl.io/master/examples.html#line-wrap-recoloring-and-scrolling). Everything else would be quite difficult I think

**Comment by FVellu** [comment]
That would be great, maybe even with links?
So you can click on a link (like this: `[Link to another page](second_file.md)`)
and it takes you to second_file.md

**Comment by minacode** [comment]
Now this is an app I want to have! Thank you @yannickulrich 🙂
I like the minimalistic style of it.

**Comment by FVellu** [comment]
Same, will this be in the next version?

**Comment by yannickulrich** [comment]
I'll make the changes to implement colour using the [syntax used in LVGL](https://docs.lvgl.io/master/widgets/label.html#text-recolor)
```
Write a #ff0000 red# word
```

Implementing links seems quite complicated as it's not directly supported by LVGL as far as I can tell. We could somehow place buttons at the correct positions or implement the touch callback. Either would be quite involved and we would have to essentially re-build the text rendering system. I don't think I understand LVGL well enough to do this but maybe someone else could give it a go..?
As it stands, this PR isn't 100% bug free and my physical watch has reset a couple of times (potentially related to copy files to the watch while the app is running). This doesn't happen in the simulator and since I have a sealed device, I can't debug this particularly well. Again, I'd appreciate any sort of input!

**Comment by Douile** [comment]
With this on watch, experience crashes when changing brightness occasionally, not sure how it causes as also have a sealed watch. App works well though and hasn't crashed while I'm using it. I feel like its todo with more stuff being in the file system making it unstable but more testing needed to confirm that.

**Comment by minacode** [comment]
I think the best way is to get this merged in a simple first version and then build additional PRs on top of it. I don't want this to be another PR dying over two years because of feature creep discussions.

**Comment by FVellu** [comment]
I'm willing to help finding bugs, how do you use your fork in InfiniSim? I replaced the InfiniTime folder with your fork but there's no gallery icon in the menu

**Comment by minacode** [comment]
The "Configure and Build" section of InfiniSims readme explains the argument you have to set. You can just link to another directory and don't have to replace something.

**Comment by FVellu** [comment]
@minacode I did build it but it still doesn't show

**Comment by minacode** [comment]
With `-DInfiniTime_DIR=path-to-your-infinitime`? I did not build it myself so I cannot verify that is does build correctly.

**Comment by FVellu** [comment]
> With `-DInfiniTime_DIR=path-to-your-infinitime`?

I think I did

File: src/displayapp/screens/ApplicationList.h [hunk 1]
```diff
@@ -36,9 +36,10 @@ namespace Pinetime {
         static constexpr int appsPerScreen = 6;
 
         // Increment this when more space is needed
-        static constexpr int nScreens = 2;
+        static constexpr int nScreens = 3;
 
         static constexpr std::array<Tile::Applications, appsPerScreen * nScreens> applications {{
+          {Symbols::gallery, Apps::Gallery},
```
Comment (minacode) [inline src/displayapp/screens/ApplicationList.h:42]: Maybe don't put this in the first slot. I suggest the slot of the Motion app which in turn gets pushed to the end of the list.

File: src/displayapp/screens/Gallery.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      class Gallery : public Screen {
+      public:
+        static constexpr const char* directory = "/gallery/";
+
+        Gallery(DisplayApp* app, Pinetime::Controllers::FS& filesystem);
+        ~Gallery() override;
+        bool OnTouchEvent(Pinetime::Applications::TouchEvents event) override;
+
+      private:
+        int string_ends_with(const char* str, const char* suffix);
```
Comment (minacode) [inline src/displayapp/screens/Gallery.h:20]: InfiniTime uses CamelCase for function names, so `StringEndsWith` as well as `ListDir` and `Open`.

**Review by minacode — CHANGES_REQUESTED** [review]
I found a few things from scrolling through your changes. This is not a full review.

File: src/displayapp/screens/Gallery.h (hunk 2 above)
Comment (yannickulrich) [reply src/displayapp/screens/Gallery.h]: Fixed with e6b5a159

File: src/displayapp/screens/ApplicationList.h (hunk 1 above)
Comment (yannickulrich) [reply src/displayapp/screens/ApplicationList.h]: Sorry, I did this mostly for testing actually. I'm happy to put it where you suggest so that it's
```cpp
          {Symbols::stopWatch, Apps::StopWatch},
          {Symbols::clock, Apps::Alarm},
          {Symbols::hourGlass, Apps::Timer},
          {Symbols::shoe, Apps::Steps},
          {Symbols::heartBeat, Apps::HeartRate},
          {Symbols::music, Apps::Music},

          {Symbols::paintbrush, Apps::Paint},
          {Symbols::paddle, Apps::Paddle},
          {"2", Apps::Twos},
          {Symbols::gallery, Apps::Gallery},
          {Symbols::drum, Apps::Metronome},
          {Symbols::map, Apps::Navigation},
          {Symbols::chartLine, Apps::Motion},
```
Fixed this with 2bef804b

**Comment by yannickulrich** [comment]
@minacode, I've made the requested changes, thanks!

Regarding colouring, we can do this trivially by
```patch
diff --git a/src/displayapp/screens/FileView.cpp b/src/displayapp/screens/FileView.cpp
index 5026c50e..7d0e3fed 100644
--- a/src/displayapp/screens/FileView.cpp
+++ b/src/displayapp/screens/FileView.cpp
@@ -97,6 +97,7 @@ TextView::TextView(uint8_t screenID, uint8_t nScreens, DisplayApp* app, const ch
   lv_obj_t* label = lv_label_create(lv_scr_act(), nullptr);
   lv_label_set_long_mode(label, LV_LABEL_LONG_BREAK);
   lv_obj_set_width(label, LV_HOR_RES);
+  lv_label_set_recolor(label, true);
   lv_obj_align(label, lv_scr_act(), LV_ALIGN_CENTER, 0, 0);

   lfs_info info = {0};
```
However, now we can't have a # in the text because LVGL is trying to parse it. We could try to engineer around this, but in the interest of getting this merged quickly (https://github.com/InfiniTimeOrg/InfiniTime/pull/1384#issuecomment-1287786186) it might be a good idea to postpone this?

**Comment by minacode** [comment]
I guess one would be more annoyed by a missing # than missing colors.

File: src/displayapp/screens/Gallery.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  } else if (StringEndsWith(fullname, ".txt")) {
+    current = std::make_unique<TextView>(n, nScreens, app, fullname, filesystem);
+  } else {
+    return false;
+  }
+
+  return true;
+}
+
+int Gallery::StringEndsWith(const char* str, const char* suffix) {
```
Comment (minacode) [inline src/displayapp/screens/Gallery.cpp:100]: This looks like it belongs to some `util` file and have a more generic name. But maybe it is fine for now.

File: src/displayapp/screens/FileView.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    c = path;
+
+  strncpy(name, c, LFS_NAME_MAX - 1);
+  char* pchar = strchr(name, '_');
+  while (pchar != nullptr) {
+    *pchar = ' ';
+    pchar = strchr(pchar + 1, '_');
+  }
+
+  pageIndicatorBasePoints[0].x = 0;
```
Comment (minacode) [inline src/displayapp/screens/FileView.cpp:25]: You are copying `displayapp/widgets/PageIndicator.cpp` here, right? Can you use it instead? Or can you modify it reasonably to use it?

**Review by minacode — CHANGES_REQUESTED** [review]
Ok, so I reviewed it. Sorry for the waiting. I am relatively new to C++ and have not done a lot with LVGL so I cannot really say something about the little details in your implementation. I seems fine to me.
But on a more conceptual level I think a separate widget for the horizontal scroll bar, or an extension of the existing vertical scrollbar, would be really nice. I am sure other PRs in the future would profit from this.

**Comment by JackRaymondCyber** [comment]
I added this app to my pinetime and it's really cool! Some quick feedback though:
- The animation for swiping between images takes a long time on the pinetime when compared to Infinisim
- Adding some text files will inexplicably crash the device causing it to restart. I don't know what causes the crash, but certain files will crash every time while others will not crash the device. Simply adding or removing a word can make the device crash/ no longer crash.
- I would personally prefer if the filenames were hidden by default and I could long press to see them. This is just my personal preference though. 

Thank you for your work!

**Comment by FVellu** [comment]
@JackRaymondCyber how did you managed to try the app?

Also, will this be included in the next release?

**Comment by JackRaymondCyber** [comment]
> @JackRaymondCyber how did you managed to try the app?
@FVellu 
I merged the changes with the base repo and then followed the build instructions. I've just made my own fork with some extra apps if you would like to try it:  https://github.com/JackRaymondCyber/InfiniTimeExtra

File: src/displayapp/screens/Gallery.cpp (hunk 3 above)
Comment (yannickulrich) [reply src/displayapp/screens/Gallery.cpp:100]: I agree with this but I don't think that we currently have `util` file. Of course I could create one but I wouldn't want to disrupt any plans the maintainers might have. I would argue that this PR should be as minimal as possible, even if that means that code needs to be moved around later on.

**Comment by yannickulrich** [comment]
Addressing @JackRaymondCyber's comments (thanks!)

> * The animation for swiping between images takes a long time on the pinetime when compared to Infinisim

This could potentially be improved by [caching the images in LVGL](https://docs.lvgl.io/latest/en/html/overview/image.html#image-caching) though am not sure.

> * Adding some text files will inexplicably crash the device causing it to restart. I don't know what causes the crash, but certain files will crash every time while others will not crash the device. Simply adding or removing a word can make the device crash/ no longer crash. This may also be a bug with ITD or or Infinitime.

Yes, unfortunately I've also experienced this. No idea why though :confused:. Since my device is sealed, I can't debug this efficiently. #560 should be able to help with this but I couldn't get this work either. If someone has an unsealed device and would be willing to poke around a bit, I'd appreciate it greatly! 

> * I would personally prefer if the filenames were hidden by default and I could long press to see them. This is just my personal preference though.

I don't have any strong opinions either. You could get this behaviour with the following patch
```patch
diff --git a/src/displayapp/screens/FileView.cpp b/src/displayapp/screens/FileView.cpp
index ec1e9d8e..734baa9c 100644
--- a/src/displayapp/screens/FileView.cpp
+++ b/src/displayapp/screens/FileView.cpp
@@ -62,7 +62,6 @@ ImageView::ImageView(uint8_t screenID, uint8_t nScreens, DisplayApp* app, const
   lv_img_set_src(image, path);
   lv_obj_align(image, lv_scr_act(), LV_ALIGN_CENTER, 0, 0);

-  ShowInfo();
 }

 TextView::TextView(uint8_t screenID, uint8_t nScreens, DisplayApp* app, const char* path, Pinetime::Controllers::FS& fs)
@@ -98,5 +97,4 @@ TextView::TextView(uint8_t screenID, uint8_t nScreens, DisplayApp* app, const ch

   fs.FileClose(&fp);

-  ShowInfo();
 }
```

**Comment by FVellu** [comment]
Will this be merged any time soon?

**Comment by JF002** [comment]
> The animation for swiping between images takes a long time on the pinetime when compared to Infinisim

This is probably caused by the relatively slow SPI bus that is shared with the external memory and the LCD. There's not much we can do here, except trying to keep the files as small as possible.

> Will this be merged any time soon?

It seems that there's at least one crash issue with this app that will need to be diagnosed and fixed before we merge the feature. Then, it'll depend on the availability of us maintainer to review and eventually merge the branch.

File: src/displayapp/screens/FileView.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  lfs_file_t fp;
+  if (fs.FileOpen(&fp, path + 2, LFS_O_RDONLY) != LFS_ERR_OK) {
+    lv_label_set_text_static(label, "could not open file");
+    lv_mem_free(buf);
+    return;
+  }
+
+  fs.FileRead(&fp, reinterpret_cast<uint8_t*>(buf), info.size);
+  lv_label_set_text(label, buf);
+  lv_mem_free(buf);
```
Comment (minacode) [inline src/displayapp/screens/FileView.cpp:97]: If I get it correctly, then `lv_label_set_text` allocates memory and copies `buf`. This uses twice the memory before `buf` is freed.
You could instead use `lv_label_set_text_static` and keep `buf alive`.

**Review by minacode — COMMENTED** [review]
The crash happens when loading specific text, right? The most interesting that happens with the loaded text is `lv_label_set_text`.

File: src/displayapp/screens/FileView.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+TextView::TextView(uint8_t screenID, uint8_t nScreens, DisplayApp* app, const char* path, Pinetime::Controllers::FS& fs)
+  : FileView(screenID, nScreens, app, path) {
+
+  lv_obj_t* label = lv_label_create(lv_scr_act(), nullptr);
+  lv_label_set_long_mode(label, LV_LABEL_LONG_BREAK);
+  lv_obj_set_width(label, LV_HOR_RES);
+  lv_obj_align(label, lv_scr_act(), LV_ALIGN_CENTER, 0, 0);
+
+  lfs_info info = {0};
+  if (fs.Stat(path + 2, &info) != LFS_ERR_OK && info.type != LFS_TYPE_DIR) {
```
Comment (minacode) [inline src/displayapp/screens/FileView.cpp:77]: Can you explain this line?

**Comment by Danimations** [comment]
Thanks for picking up this concept and running with it. I'm looking forward to seeing it evolve over time.

File: src/displayapp/screens/Gallery.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  if (res != 0) {
+    NRF_LOG_INFO("[Gallery] can't find directory");
+    return;
+  }
+  while (filesystem.DirRead(&dir, &info)) {
+    if (info.type == LFS_TYPE_DIR)
+      continue;
+    nScreens++;
+  }
+  assert(filesystem.DirClose(&dir) == 0);
```
Comment (yannickulrich) [inline src/displayapp/screens/Gallery.cpp:53]: I investigated the issue a bit further and have the feeling that the issue might be related to LittleFS. I have commented out almost all functionality and am just running this `DirRead` loop. Everything else, including the creation of the actual screens in `Gallery::Open`, is commented out. This still crashes on my watch when I do the following
 * open the gallery app
 * ignore the fact that it's completely empty
 * exit the gallery app,
 * go to any setting (eg. Time format)
 * try to change it and exit setting

To confirm it's related to the `DirRead` and nothing else, I then hardcoded a file path in `Gallery::Open` and disable the `DirRead` completely. In that case the bug did not appear. So I'd guess it's related to this loop

File: src/displayapp/screens/FileView.cpp (hunk 6 above)
Comment (yannickulrich) [reply src/displayapp/screens/FileView.cpp:77]: you mean the `path+2` bit? This is because `Gallery::Open` passes the path as eg. `F:/gallery/Message.txt`. However, LittleFS only accepts `/gallery/Message.txt` so I add 2 to the pointer to drop the first two chars.

It's a bit annoying that LVGL must have the `F:` and LittleFS must not. Of course we could *add* the `F:` in `ImageView::ImageView` but that'd require more low-level memory operations than just `+2`

File: src/displayapp/screens/FileView.cpp (hunk 5 above)
Comment (yannickulrich) [reply src/displayapp/screens/FileView.cpp:97]: You'd suggest freeing `buf` in a `TextView::~TextView`? We could do that, sure.

**Review by yannickulrich — COMMENTED** [review]
Sorry for the late response.. I did some investigating (see above)

File: src/displayapp/screens/FileView.cpp (hunk 6 above)
Comment (minacode) [reply src/displayapp/screens/FileView.cpp:77]: No, I meant the meaning of the whole boolean expression. I could not get what it does.

File: src/displayapp/screens/FileView.cpp (hunk 5 above)
Comment (yannickulrich) [reply src/displayapp/screens/FileView.cpp]: Fixed in 95af87f0

File: src/displayapp/screens/FileView.cpp (hunk 6 above)
Comment (yannickulrich) [reply src/displayapp/screens/FileView.cpp]: You're right, it's wrong as it is. The idea was to check if the `Stat` call worked and then if the file is a regular file and not a directory. I'm splitting it into two ifs now 8e0a13f7

**Comment by Douile** [comment]
> The animation for swiping between images takes a long time on the pinetime when compared to Infinisim

Using `DisplayApp::FullRefreshDirections::None` seems to make images render a lot faster at the cost of not having a pretty swipe animation (Tested on PineTime).

File: src/displayapp/screens/Gallery.cpp (hunk 7 above)
Comment (IchbinkeinReh) [reply src/displayapp/screens/Gallery.cpp:53]: I think i just found the issue. You are calling DirClose inside the assert which you should not do. If your move the call to a seperate variable then the issue seems to disappear. My watch ran pretty stable after i changed it

**Comment by IchbinkeinReh** [comment]
@yannickulrich Will you update the PR and resolve merge conflicts? Otherwise i would be willing to open a new PR with resolved merge conflicts if you are not interested in fixing this PR anymore.

**Comment by IchbinkeinReh** [comment]
Also a note for everyone that is desperately trying to upload resource files to the watch. I found https://infinitimeexplorer.netlify.app/ to work best for uploading the files. It worked flawlessly from my android phone.

File: src/displayapp/screens/Gallery.cpp (hunk 7 above)
Comment (yannickulrich) [reply src/displayapp/screens/Gallery.cpp:53]: Very nice, thanks for figuring it out :partying_face:  I've been using a modified version of the PR for a while now that circumvented the `DirRead` problem but was quite annoying to use. I'll fix this asap

**Comment by yannickulrich** [comment]
> @yannickulrich Will you update the PR and resolve merge conflicts? Otherwise i would be willing to open a new PR with resolved merge conflicts if you are not interested in fixing this PR anymore.

I'll update this later today, thanks for the offer though :blush: 

For those of you using itd to connect to your watch, I have opened a [pull request](https://gitea.arsenm.dev/Arsen6331/itd/pulls/55) over there as well to make using this app easier by mounting the filesystem directly into your computer

File: src/displayapp/screens/Gallery.cpp (hunk 7 above)
Comment (yannickulrich) [reply src/displayapp/screens/Gallery.cpp]: fixed this in 15bce7f5

**Comment by yannickulrich** [comment]
I have rebased on the current main branch with all fixes in place. The main one is of course the crash (thanks, @IchbinkeinReh) but I've also removed the swipe animation (thanks, @Douile. we still get a bit of animation but it's much easier to access barcodes in a pinch). Let me know what you think.

**Comment by tigercoding56** [comment]
now that you are talking about markdown , why not have html support?

**Comment by minacode** [comment]
Please don't overengineer this PR to death 😄

**Comment by yannickulrich** [comment]
I have rebased everything to `main` and fixed a formatting bug. I'm not entirely sure how to continue but want to point out that all bugs we are aware of have been fixed. Can someone please advise how to move forward from this (eg. @JF002)?

**Comment by Findirfin** [comment]
I would very much appreciate if this came out in the next version. I'm loving infinitime, but this would make it way more useful for me.

**Comment by ProgramminCat** [comment]
Please merge because adding a gallery app to InfiniTime, the open source FreeRTOS-based OS for smartwatches, could benefit the community in the following ways:

    More functionality - A gallery app would allow users to view photos stored on the watch, making the watch more useful as a standalone device. This adds a commonly desired feature.
    Showcase capabilities - By demonstrating features like accessing local storage and displaying images/media, a gallery app would showcase InfiniTime's capabilities as a smartwatch OS. It validates it as a practical alternative.
    Attract new users - A gallery, along with other apps, makes the OS more attractive for new users looking for an open smartwatch platform. It improves adoption.
    Contribute to ecosystem - Being open source, a gallery app contribution would grow the app ecosystem and encourage more developers to build for InfiniTime.
    Feedback for improvements - Users would provide feedback on the gallery app which could lead to fixes and additions that improve the overall OS.
    Customization options - Developers could build alternate gallery apps with different features catering to customization needs.
    Security enhancements - Handling access to local storage securely in the gallery app could lead to hardened OS-level security enhancements.

So in summary, a contributed gallery app would bring a commonly expected feature, demonstrate InfiniTime's capabilities, attract new users, grow the dev ecosystem, enable improvements, support customizations, and potentially improve security - benefitting the entire open source community.

**Comment by minacode** [comment]
I don't think that this PR isn't merged because of a lack of good arguments.

**Comment by yannickulrich** [comment]
@minacode, do you have any idea what is holding this up? Are the maintainers just overworked or is there a specific problem I can address?

**Comment by minacode** [comment]
No, sadly not. I guess it's a time issue. See how many PRs are open and get merged by a few people (thanks 🙏🏼). Currently there is a bug issue from the 1.13 release that will surely get a lot of the attention. Apart from that there is only little memory left, so merges have to be done wisely. I am tinkering with optional apps for some time now to hopefully improve this someday. I would like to see my calculator get merged as well and I think the best way to get there is to improve the whole situation around apps.

**Comment by minacode** [comment]
Update: #1894 was merged 🥳

**Comment by yannickulrich** [comment]
In case anyone's interested, I have rebased this again onto main in case some of the maintainers are interested into getting this merged (esp. since https://github.com/InfiniTimeOrg/InfiniTime/pull/1894 removes the danger of running out of space). If anyone has comments, please let me know 😊

**Comment by mark9064** [comment]
Tentatively adding this to the milestone :) will try and review when I have time

**Comment by liamcharger** [comment]
It's worth noting that [this issue](https://github.com/InfiniTimeOrg/InfiniTime/issues/2021) will be very prominent with this app since it lists dirs and reads files often, as confirmed in this comment: https://github.com/InfiniTimeOrg/InfiniTime/pull/1384#issuecomment-1287777881

**Comment by mark9064** [comment]
Is that linked issue specific to access over BLE or does it apply generally?

**Comment by liamcharger** [comment]
Hm good question..

I imagine it's in general, especially after someone mentioned it while using this app, but it would be a good idea to double check
